parser/prism/Translator.cc

```diff
@@ -1193,27 +1207,45 @@ std::unique_ptr<parser::Node> Translator::translateStatements(pm_statements_node
 
-    std::string_view name = parser.resolveConstant(node->name);
+    auto constexpr isConstantPath = std::is_same_v<PrismLhsNode, pm_constant_path_target_node> ||
+                                    std::is_same_v<PrismLhsNode, pm_constant_path_write_node> ||
+                                    std::is_same_v<PrismLhsNode, pm_constant_path_node>;
 
     std::unique_ptr<parser::Node> parent;
-    if (node->parent) {
-        // This constant reference is chained onto another constant reference.
-        // E.g. if `node` is pointing to `B`, then then `A` is the `parent` in `A::B::C`.
-        parent = translate(node->parent);
-    } else {                                                // This is a fully qualified constant reference, like `::A`.
-        pm_location_t *delimiterLoc = &node->delimiter_loc; // The location of the `::`
-        parent = make_unique<parser::Cbase>(parser.translateLocation(delimiterLoc));
+    if constexpr (isConstantPath) { // Handle constant paths, has a parent node that needs translation.
+        if (auto prismParentNode = node->parent; prismParentNode != nullptr) {
+            // This constant reference is chained onto another constant reference.
+            // E.g. given `A::B::C`, if `node` is pointing to the root, `A::B` is the `parent`, and `C` is the `name`.
+            //   A::B::C
+            //    /    \
+            //  A::B   ::C
+            //  /  \
+            // A   ::B
+            parent = translate(prismParentNode);
+        } else { // This is the root of a fully qualified constant reference, like `::A`.
+            pm_location_t *delimiterLoc = &node->delimiter_loc; // The location of the `::`
+            parent = make_unique<parser::Cbase>(parser.translateLocation(delimiterLoc));
+        }
+    } else { // Handle plain constants like `A`, that aren't part of a constant path.
+        static_assert(std::is_same_v<PrismLhsNode, pm_constant_and_write_node> ||
+                      std::is_same_v<PrismLhsNode, pm_constant_or_write_node> ||
+                      std::is_same_v<PrismLhsNode, pm_constant_operator_write_node> ||
+                      std::is_same_v<PrismLhsNode, pm_constant_target_node> ||
+                      std::is_same_v<PrismLhsNode, pm_constant_read_node> ||
+                      std::is_same_v<PrismLhsNode, pm_constant_write_node>);
+        parent = nullptr;
     }
 
-    if (isAssignment) {
-        return make_unique<parser::ConstLhs>(parser.translateLocation(loc), std::move(parent),
-                                             gs.enterNameConstant(name));
-    } else {
-        return make_unique<parser::Const>(parser.translateLocation(loc), std::move(parent), gs.enterNameConstant(name));
-    }
+    pm_location_t *loc = &node->base.location;
+    std::string_view name = parser.resolveConstant(node->name);
+
+    return make_unique<SorbetLHSNode>(parser.translateLocation(loc), std::move(parent), gs.enterNameConstant(name));
```

Comment:
This is always either `make_unique<Const>` or `make_unique<ConstLhs>`.

Sorbet handles constants and constant paths the same way, just that the former has the parent as `nullptr` (hence line 1242 above)